physics/solar_system_body.hpp

```diff
@@ -337,9 +394,72 @@ template<typename Frame>
 std::vector<DegreesOfFreedom<Frame>>
 SolarSystem<Frame>::MakeAllDegreesOfFreedom() {
   std::vector<DegreesOfFreedom<Frame>> degrees_of_freedom;
-  for (auto const& pair : initial_state_map_) {
-    serialization::InitialState::Body const* const body = pair.second;
-    degrees_of_freedom.push_back(MakeDegreesOfFreedom(*body));
+  if (!cartesian_initial_state_map_.empty()) {
+    for (auto const& pair : cartesian_initial_state_map_) {
+      serialization::InitialState::Cartesian::Body const* const body =
+          pair.second;
+      degrees_of_freedom.push_back(MakeDegreesOfFreedom(*body));
+    }
+  }
+  if (!keplerian_initial_state_map_.empty()) {
+    // First, construct all the bodies and find the primary body of the system.
+    std::string primary;
+    std::map<std::string,
+             not_null<std::unique_ptr<MassiveBody const>>> owned_bodies;
+    std::map<std::string, not_null<MassiveBody const*>> unowned_bodies;
+    for (auto const& pair : keplerian_initial_state_map_) {
+      const auto& name = pair.first;
+      serialization::InitialState::Keplerian::Body const* const body =
+          pair.second;
+      CHECK_EQ(body->has_parent(), body->has_elements()) << name;
+      if (!body->has_parent()) {
+        CHECK(primary.empty()) << name;
+        primary = name;
+      }
+      auto owned_body = MakeMassiveBody(gravity_model_message(name));
+      unowned_bodies.emplace(name, owned_body.get());
+      owned_bodies.emplace(name, std::move(owned_body));
+    }
+
+    // Construct a hierarchical system rooted at the primary and add the other
+    // bodies layer by layer.
```

Comment:
This seems a bit complicated; if I recall correctly `HierarchicalSystem` only requires that a body be added after its parent, and I am fine with requiring that the proto refer to a body only after it has been defined (that seems like good hygiene anyway).